src/hotspot/share/gc/shenandoah/shenandoahLock.cpp

```diff
@@ -28,9 +28,58 @@
 
 #include "gc/shenandoah/shenandoahLock.hpp"
 #include "runtime/atomic.hpp"
+#include "runtime/interfaceSupport.inline.hpp"
 #include "runtime/javaThread.hpp"
 #include "runtime/os.inline.hpp"
 
+// These are inline variants of Thread::SpinAcquire with optional blocking in VM.
+// Unfortunately, there is no good way to do a conditional RAII object in the
+// middle of the method without dealing with macros.
+
+void ShenandoahLock::contended_lock(bool allow_block_for_safepoint) {
+  Thread* thread = Thread::current();
+  if (allow_block_for_safepoint && thread->is_Java_thread()) {
+    contended_lock_or_block(JavaThread::cast(thread));
+  } else {
+    contended_lock_no_block();
+  }
+}
+
+void ShenandoahLock::contended_lock_no_block() {
+  int ctr = 0;
+  int yields = 0;
+  while (Atomic::cmpxchg(&_state, 0, 1) != 0) {
+    if ((++ctr & 0xFFF) == 0) {
+      if (yields > 5) {
+        os::naked_short_sleep(1);
+      } else {
+        os::naked_yield();
+        yields++;
+      }
+    } else {
+      SpinPause();
+    }
+  }
+}
+
+void ShenandoahLock::contended_lock_or_block(JavaThread* java_thread) {
```

Comment:
As this is a copy you could use ThreadBlockInVM or a NoOp as template parameter to get one version.

Reply:
Right, that's a good idea. Let me see...

Reply:
Done in new commit. Retesting...